### PR TITLE
[backport] Make `-target` support JDK 8 through 19 (and deprecate 5 through 7)

### PR DESCRIPTION
**Pull request by stefan-jurco**
Even though [scala >= 2.12.15 supports jvm >=8](https://docs.scala-lang.org/overviews/jdk-compatibility/overview.html). It is not possible to use it as a target. This adds `jvm-11`, `jvm-17` and `jvm-18` alongside `jvm-1.8` as supported target values.

### CODE REVIEW

**Comment by SethTisue** [comment]
Is this a backport from 2.13?

**Comment by stefan-jurco** [comment]
Just the test `TargetTest.scala` as the `-target` resolution is completely changed in `2.13` and I wanted the change to be as minimal as possible.

**Comment by SethTisue** [comment]
I agree that it's fine to pretend 9,10,12,13,14,15,16 don't exist.

It seems peculiar to me to require the new "jvm-11" style for the new flags, but then not to also support "jvm-8". If "jvm-8" actually works, could you test that in `TargetTests`? And if it isn't supported, could you add the support?

Have you verified that the bytecode version actually gets set as expected?

**Comment by stefan-jurco** [comment]
Backported support for all jvm versions as are in `2.13.x` branch with the same alternative formats.
5,6,7 versions are still deprecated.

The bytecode version was not set, I missed that in the code, thanks. And it's verified but just locally as I haven't found test for that.

```bash
$ for v in `seq 5 18`; do scalac -deprecation -target:$v Test.scala; file Test.class; done
warning: -target is deprecated: -target:5 is deprecated, forcing use of 8
one warning found
Test.class: compiled Java class data, version 52.0 (Java 1.8)
warning: -target is deprecated: -target:6 is deprecated, forcing use of 8
one warning found
Test.class: compiled Java class data, version 52.0 (Java 1.8)
warning: -target is deprecated: -target:7 is deprecated, forcing use of 8
one warning found
Test.class: compiled Java class data, version 52.0 (Java 1.8)
Test.class: compiled Java class data, version 52.0 (Java 1.8)
Test.class: compiled Java class data, version 53.0 (Java SE 9)
Test.class: compiled Java class data, version 54.0 (Java SE 10)
Test.class: compiled Java class data, version 55.0 (Java SE 11)
Test.class: compiled Java class data, version 56.0 (Java SE 12)
Test.class: compiled Java class data, version 57.0 (Java SE 13)
Test.class: compiled Java class data, version 58.0 (Java SE 14)
Test.class: compiled Java class data, version 59.0
Test.class: compiled Java class data, version 60.0
Test.class: compiled Java class data, version 61.0
Test.class: compiled Java class data, version 62.0
```

**Comment by stefan-jurco** [comment]
Hi @SethTisue is there anything else I should update?

**Review by SethTisue — COMMENTED** [review]
lgtm... maybe @som-snytt would like to have a look?

File: project/ScalaOptionParser.scala [hunk 1]
```diff
@@ -126,4 +126,5 @@ object ScalaOptionParser {
   private def scaladocPathSettingNames = List("-doc-root-content", "-diagrams-dot-path")
   private def scaladocMultiStringSettingNames = List("-doc-external-doc")
 
+  private val targetSettingNames = (5 to 18).map(_.toString).flatMap(v => v :: s"jvm-1.$v" :: s"jvm-$v" :: s"1.$v" :: Nil).toList
```
Comment (som-snytt) [inline project/ScalaOptionParser.scala]:
```suggestion
  private val targetSettingNames = (5 to 18).flatMap(v => s"$v" :: s"jvm-1.$v" :: s"jvm-$v" :: s"1.$v" :: Nil).toList
```

Comment (stefan-jurco) [reply project/ScalaOptionParser.scala]: updated

**Comment by stefan-jurco** [comment]
@som-snytt @SethTisue do you want to review it again or could it be merged?

**Review by SethTisue — APPROVED** [review]

**Comment by som-snytt** [comment]
I was like what's a `ChoiceSettingForcedDefault`? I still don't know, but at least now I realize this is a backport to 2.12, not to 2.13.

LGTM within an approximation.

(Noting "wants to merge ... into `scala:2.12.x`" on the github UI is in tiny glowing blue font on my screen which is not a phone. I'm wearing the correct eyeglasses.)

OK, just read the PR on 2.13 https://github.com/scala/scala/pull/8060

I would have said the preHook was jumping the shark. I sense a peter pan pun. Oh that was an alligator not a shark. Or a crocodile.

I was aware that `AbstractSettings` went away at some point, and also that those classes are frustrating to work with, as expressed on that PR.

Footnote: I would have gone with just listing the variants, `jvm-1.8`, `8`, etc. and definitely not supported `jvm-17`. Where did I just see someone's joking indignation at `1.9`? `ScalaVersion` is the type for versions, but it doesn't parse arbitrary prefixes.

**Review by som-snytt — APPROVED** [review]

**Comment by SethTisue** [comment]
#10000 extends this to JDK 19.

**Pull request merged**